### PR TITLE
Sections support

### LINKED ISSUE #28182: Sections support in Blazor
Basically the same as https://github.com/dotnet/aspnetcore/issues/10131 which has been closed in in favour of #10452 which was in turn closed in favour of #10450 (influencing the `<head>`). 

Having the ability to embed content from the page in the layout in more than one placeholder is a common need. A typical use case is to display the page name in the "header" of the application which is usually defined in the layout. Another use case is to display personalised content for the currently logged user.

A lot of other technologies support this feature and I think it would be a great addition to Blazor. A few examples:

- ASP.NET WebForms has [ContentPlaceholder](https://docs.microsoft.com/en-us/aspnet/web-forms/overview/older-versions-getting-started/master-pages/multiple-contentplaceholders-and-default-content-cs)
- ASP.NET MVC has [sections](https://docs.microsoft.com/en-us/aspnet/core/mvc/views/layout?view=aspnetcore-5.0)
- Web Pages also have sections
- Angular has [router outlets](https://angular.io/api/router/RouterOutlet#description)
- React Router also [supports multiple components](https://gist.github.com/kitdemonio/33be59d4ab597f2679e83d68e1b63842)

Thank you for your consideration!

### PR DESCRIPTION
**Pull request by surayya-MS**
# Sections support

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

API Proposal: https://github.com/dotnet/aspnetcore/issues/46937

## Description

* Made `SectionContent` and `SectionOutlet` public
* Removed `SectionContent.IsDefaultContent`
* Removed `Name` and used object `SectionId` as a section identifier
* Added E2E tests for Sections

Fixes #28182

### CODE REVIEW

File: src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj
Comment (MackinnonBuck) [inline src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj:52]: Could we undo changes to this file? Just to keep its git history clean 🙂

File: src/Components/Web/src/Head/HeadOutlet.cs
Comment (MackinnonBuck) [inline src/Components/Web/src/Head/HeadOutlet.cs:18]: This is good 👍

I do wonder if we could go even further and replace the `Name` parameter on `SectionContent` and `SectionOutlet` with an `object` parameter, and call it `Id`? This would enable libraries to be completely certain they're not picking a `Name` that conflicts with another library (or the user's app).

You could do something like:

`MyLibraryComponent.razor`:
```razor
<h1>This is a component in an RCL<h1/>

@*The app can inject additional content here*@
<SectionOutlet Id="SectionOutletId">

@code {
    public static readonly object SectionOutletId = new();

    // ...
}
```

`MyAppComponent.razor`
```razor
<h1>This is a component in my app<h1/>

@*Inject some content into the <MyLibraryComponent /> somewhere else in the app*@
<SectionOutlet Id="MyLibraryComponent.SectionOutletId">
    <span>Here is some content</span>
</SectionOutlet>
```

This would also enable the scenario where there are potentially multiple instances of a component rendering a `<SectionOutlet />`. You could make the `SectionOutletId` non-static and let other components pick which "instance" to render to (via a cascading value, DI service, etc.).

Do you/others have any thoughts on this? @dotnet/aspnet-blazor-eng

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (MackinnonBuck) [inline src/Components/test/E2ETest/Tests/SectionsTest.cs:35]: We have `Browser.Exists()` for this

File: src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponent.razor
Comment (MackinnonBuck) [inline src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponent.razor:13]: Nit: We usually put `@code` blocks at the bottom of the file

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (MackinnonBuck) [inline src/Components/test/E2ETest/Tests/SectionsTest.cs:13]: I can think of a few additional scenarios that might be worth testing:
* Multiple `SectionContent` instances render to the same `SectionOutlet`:
  * If `IsDefaultContent` is `false` for a `SectionContent`, it will override existing content in its corresponding `SectionOutlet`.
  * If `DefaultContent` is `true`, it will not replace existing content in the `SectionOutlet`, but its content _will_ get rendered if all previously-registered `SectionContent`s (including those with `DefaultContent == true`) are disposed.
* The `Name` parameter of a `SectionContent` instance gets changed, causing it to match a different `SectionOutlet`.
* The `IsDefaultContent` parameter of a `SectionContent` gets changed (I actually don't think we handle this case currently... should we?)
* A `SectionContent` attempts to render to a `SectionOutlet` that doesn't exist.
  * The `SectionOutlet` never existed.
  * The `SectionOutlet` gets disposed before the `SectionContent` does.
* All `SectionContent` instances rendering to a `SectionOutlet` get disposed.

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (Nick-Stanton) [inline src/Components/test/E2ETest/Tests/SectionsTest.cs:27]: I just found out while working on E2E tests for QuickGrid that `noReload` was removed in https://github.com/dotnet/aspnetcore/pull/35414 and setting it doesn't have any effect. @SteveSandersonMS was there a reason to keep it as an argument for `WebDriverExtensions.Navigate()`?

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (surayya-MS) [reply src/Components/test/E2ETest/Tests/SectionsTest.cs:13]: >The IsDefaultContent parameter of a SectionContent gets changed (I actually don't think we handle this case currently)

Yes, we currently don't handle this case. I'm working on this.

I covered all these cases.

Could you please clarify what is expected to happen in case when two `SectionContent` instances with the same `Name` are set as default (`IsDefaultContent=true`) ?

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (surayya-MS) [reply src/Components/test/E2ETest/Tests/SectionsTest.cs:27]: @Nick-Stanton thanks! I'll investigate this.

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (MackinnonBuck) [reply src/Components/test/E2ETest/Tests/SectionsTest.cs:13]: > Could you please clarify what is expected to happen in case when two `SectionContent` instances with the same `Name` are set as default (`IsDefaultContent=true`) ?

The way the current implementation was designed to work, the earliest-initialized `SectionContent` with `IsDefaultContent=true` should be rendered with higher priority. Or, in other words, the last-initialized default `SectionContent` is "more default" than the previous default `SectionContent`, in the sense that it's treated as the default content when the previous default `SectionContent` gets disposed.

File: src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponentWithTwoChildren.razor
Comment (MackinnonBuck) [inline src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponentWithTwoChildren.razor:14]: Nit: We generally use hyphens ("-") instead of underscores ("_") for `id` and `class` HTML attribute values.

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (MackinnonBuck) [inline src/Components/test/E2ETest/Tests/SectionsTest.cs:13]: Supernit:
```suggestion
namespace Microsoft.AspNetCore.Components.E2ETests.Tests;

public class SectionsTest : ServerTestBase<ToggleExecutionModeServerFixture<Program>>
```

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (surayya-MS) [reply src/Components/test/E2ETest/Tests/SectionsTest.cs:43]: I opened an issue for this https://github.com/dotnet/aspnetcore/issues/46810

**Review by MackinnonBuck — COMMENTED** [review]
This is looking great! I just added a few more comments with some minor thoughts.

Before approving, I would still like to see what others think about [this suggestion](https://github.com/dotnet/aspnetcore/pull/46727/files#r1110151185).

File: src/Components/Web/src/Head/HeadOutlet.cs
Comment (MackinnonBuck) [reply src/Components/Web/src/Head/HeadOutlet.cs:18]: > At the same time I am fine with using string `Name` instead of object `Id` as it is a convention in Html to use string for elements ids.

Yeah, I can see how naming the component parameter `Id` could be seen as somehow relating to the `id` HTML element attribute. Perhaps using a different parameter name would avoid this problem, maybe `Identifier` or `SectionKey`?

File: src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponentWithTwoChildren.razor
Comment (surayya-MS) [reply src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponentWithTwoChildren.razor:91]: No reason. I'll make them private. Thanks!

File: src/Components/Web/src/Head/HeadOutlet.cs
Comment (surayya-MS) [reply src/Components/Web/src/Head/HeadOutlet.cs:18]: I think `SectionId` sounds good.

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (surayya-MS) [reply src/Components/test/E2ETest/Tests/SectionsTest.cs:31]: Thanks! In that case I'd better not cache them.

File: src/Components/Web/src/Head/HeadOutlet.cs
Comment (MackinnonBuck) [inline src/Components/Web/src/Head/HeadOutlet.cs:18]:
```suggestion
    internal static readonly object HeadSectionOutletId = new();
    internal static readonly object TitleSectionOutletId = new();
```

File: src/Components/Components/src/Sections/SectionOutlet.cs
Comment (surayya-MS) [reply src/Components/Components/src/Sections/SectionOutlet.cs:39]: Very good point!

Indeed this is becoming too complex. We decided to use object `SectionId` instead of string `Name` to avoid any naming issues. But, as you pointed out, the users can assign anything to `SectionId` (could be reference type in your example or could be string which is immutable and there will be equality). So now  if we decided (most likely not) to leave object `SectionId` as it is then we are relying on our users to know how reference and value types are allocated in the memory. With the current solution we cannot force them to use only object for `SectionId`. That could potentially introduce confusion for the users. We need something else. 
I like the way you think to fix equality checking but it is becoming too complex.

I think we should go back to string `Name`. It is simple and intuitive. The only downside to this is when there are two `SectionOutlet`s rendered on the same page with the same `Name` but an exception will be thrown saying that `SectionOutlet` with the same name already exists, so this is fine. Or another case when there is a `SectionContent` with some `Name` and it happens that there already a `SectionOutlet` with the same `Name` on this page. The user will see that the content is rendered and will figure out that the `SectionOutlet` had the same `Name`, so this is fine too.

What do you think?

Comment (surayya-MS) [reply src/Components/Components/src/Sections/SectionOutlet.cs:39]: > 2. You want to use value equality for a reference type
Let's say I've defined a custom MySectionIdentifier class that implements IEquatable<MySectionIdentifier>, and I plan on using an instance of this class as the SectionId for my SectionOutlet. There would currently be no way of enabling a SectionContent to render to my SectionOutlet without somehow obtaining a reference to the exact MySectionIdentifier passed to my SectionOutlet.

I thought that was the whole point to use the exact same object for `SectionId` of `SectionContent` as in `SectionOutlet`. The only way is to pass it as parameter to the child component or use static field in the parent component.

This brings me to a solution to fix the issue with casting object. We can introduce new class `SectionIdentifier` and make `SectionId` of type `SectionIdentifier`. But I think it is better to go back to string `Name`.

Comment (surayya-MS) [reply src/Components/Components/src/Sections/SectionOutlet.cs:39]: As discussed, we are implementing the design where `SectionId` is object

File: src/Components/Components/src/Sections/SectionContent.cs
Comment (MackinnonBuck) [inline src/Components/Components/src/Sections/SectionContent.cs:45]: Can we update this exception message?
```suggestion
            throw new InvalidOperationException($"{nameof(SectionContent)} requires a non-null value for the parameter '{nameof(SectionId)}'.");
```

File: src/Components/Components/src/Sections/SectionOutlet.cs
Comment (MackinnonBuck) [inline src/Components/Components/src/Sections/SectionOutlet.cs:19]: Nit:
```suggestion
    /// Gets or sets the ID that determines which <see cref="SectionContent"/> instances will provide
```

File: src/Components/Components/src/Sections/SectionOutlet.cs
Comment (MackinnonBuck) [inline src/Components/Components/src/Sections/SectionOutlet.cs:36]:
```suggestion
            throw new InvalidOperationException($"{nameof(SectionOutlet)} requires a non-null value for the parameter '{nameof(SectionId)}'.");
```

File: src/Components/Components/src/Sections/SectionRegistry.cs
Comment (MackinnonBuck) [inline src/Components/Components/src/Sections/SectionRegistry.cs:33]: For debugging purposes, it might be nice to state what the value of `sectionId` was when this failed.
```suggestion
            throw new InvalidOperationException($"There are no content providers with the given section ID '{sectionId}'.");
```
But tbh this is a minor detail, so I'll leave it up to you in case you disagree. If you don't take that change, could we at least update it to:
```suggestion
            throw new InvalidOperationException($"There are no content providers with the given {nameof(SectionId)}.");
```
That way, we don't risk forgetting to update the exception message if we rename `SectionId` in the future 🙂

Comment (MackinnonBuck) [reply src/Components/Components/src/Sections/SectionRegistry.cs:33]: Oh, and please apply changes to similar exception messages in this file to match whatever you decide to do here. Thanks!

File: src/Components/Web/src/Head/HeadOutlet.cs
Comment (MackinnonBuck) [inline src/Components/Web/src/Head/HeadOutlet.cs:18]: This is a major nit and I'm so sorry to do this again, but I just realized that these names actually make a bit more sense:
```suggestion
    internal static readonly object HeadSectionId = new();
    internal static readonly object TitleSectionId = new();
```

File: src/Components/test/E2ETest/Tests/SectionsTest.cs
Comment (MackinnonBuck) [inline src/Components/test/E2ETest/Tests/SectionsTest.cs:70]: Nit:
```suggestion
        // Render Counter and TextComponent SectionContents with same Name
        // TextComponent SectionContent overrides Counter SectionContent
```

File: src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponentWithTwoChildren.razor
Comment (MackinnonBuck) [inline src/Components/test/testassets/BasicTestApp/SectionsTest/ParentComponentWithTwoChildren.razor:86]: Nit: there is still some use of "Name" in this file when it should be "Id" now.

**Review by MackinnonBuck — APPROVED** [review]
This is looking great 🎉 Thank you for your perseverance on this!

I have a few remaining nits but nothing that should block the approval of this PR. I think this is ready to go after all tests pass (and the API gets approved).

File: src/Components/Web/src/Head/HeadOutlet.cs
Comment (surayya-MS) [reply src/Components/Web/src/Head/HeadOutlet.cs:18]: I agree

File: src/Components/Components/src/Sections/SectionRegistry.cs
Comment (surayya-MS) [reply src/Components/Components/src/Sections/SectionRegistry.cs:33]: I have thought about it. 

The first one probably won't be very helpful since `SectionId` is object. The exception would be "There are no content providers with the given section ID 'System.Object'."

With the second one the problem is that `SectionRegistry` doesn't know about `SectionOutlet` or `SectionContent`. It works with `ISectionContentSubscriber` and `ISectionContentProvider` which makes sense. There is no `SectionId` in those interfaces. 
We could use `{nameof(SectionOutlet.SectionId)}` (depending on it is subscriber or provider) which I am fine with since `SectionRegistry` is internal. 
Or we could use `{nameof(sectionId)}` the name of the variable. This would not be of much help in case we rename `SectionId` in `SectionOutlet` and `SectionContent` we still need to remember to change the variable's name in `SectionRegistry` too.

What do you think?

Comment (MackinnonBuck) [reply src/Components/Components/src/Sections/SectionRegistry.cs:33]: > "There are no content providers with the given section ID 'System.Object'."

It should still look fine if the user provides a value with a proper `ToString()` implementation (like a `string`, etc.). But I agree it can be confusing in some cases. Totally fine if you prefer to not include the value in the exception message, especially considering this is an exception that most likely indicate something incorrect about the _internal_ implementation, not the way the feature is being used. So, it's unlikely a customer will ever see it anyway (our code is perfect 😇).

> We could use {nameof(SectionOutlet.SectionId)}

Yeah, that seems good. Maybe use `SectionContent`'s instead simply because the exception message refers to a "content provider" which implies `SectionContent`.

_Or_... we could just use the phrase "section ID". That way it refers to the _concept_ of a section ID and not a literal symbol in code (the string "SectionId" implies there is some sort of connection to a variable, which is why I initially suggested using `nameof`).

Comment (surayya-MS) [reply src/Components/Components/src/Sections/SectionRegistry.cs]: >we could just use the phrase "section ID". That way it refers to the concept

I like this option

File: src/Components/Components/src/PublicAPI.Unshipped.txt
Comment (SteveSandersonMS) [inline src/Components/Components/src/PublicAPI.Unshipped.txt:11]: Do you think it would be viable to remove `IsDefaultContent` from the public API? We added it in https://github.com/dotnet/aspnetcore/pull/35982 as a workaround for a bug at that time, but we didn't worry too much about the usability of this flag since it was only used internally. With `<SectionContent>` becoming public, we probably wouldn't want to expose this flag since it will be really hard for people to make sense of.

I'm not actually sure if `IsDefaultContent` is still required now that we fixed the issues about the order in which prerendering occurs. The following comment from @MackinnonBuck may no longer apply, but I haven't checked:

> An alternative I considered was to always render the SectionContent for the default title, but only render the <title> element when the default title is actually determined. This sometimes works, but it breaks in cases like prerendering because <PageTitle> elements in the "app" root component still get rendered first.

If it's possible to remove `IsDefaultContent` completely (rather than just making it `internal`) that would be the best outcome but I'm not 100% sure if that's possible! Is that something you'd be able to check, @surayya-MS?

Comment (surayya-MS) [reply src/Components/Components/src/PublicAPI.Unshipped.txt:11]: Sure, we could make `IDefaultContent` `internal`. I cannot answer now if we could remove it completely. It is used in [HeadOutlet](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web/src/Head/HeadOutlet.cs#L48). I would have to investigate that.

>I'm not actually sure if IsDefaultContent is still required now that we fixed the issues about the order in which prerendering occurs.

@SteveSandersonMS could you please point out the PRs where we fixed the issues about the order in which prerendering occurs?

@MackinnonBuck , do you think it is possible to remove `ISDefaultContent` completely?

Comment (SteveSandersonMS) [reply src/Components/Components/src/PublicAPI.Unshipped.txt:11]: Yes, the rendering order issues were fixed in https://github.com/dotnet/aspnetcore/pull/40512, which was after Mackinnon implemented the `<head>` support in https://github.com/dotnet/aspnetcore/pull/34503.

Comment (surayya-MS) [reply src/Components/Components/src/PublicAPI.Unshipped.txt:11]: I removed `IsDefaultContent` and tried out the bug described in the PR https://github.com/dotnet/aspnetcore/pull/35982 . It is no longer an issue. 
Btw it wouldn't be possible to make `IsDefaultContent` `internal` because it has `Parameter` attribute.

Comment (SteveSandersonMS) [reply src/Components/Components/src/PublicAPI.Unshipped.txt:11]: > It is no longer an issue

That's great news!

> It wouldn't be possible to make IsDefaultContent internal because it has Parameter attribute.

There might still be ways to achieve it (by hand-writing the logic that applies parameters in `SetParametersAsync` instead of auto-assigning them based on `[Parameter]` attributes) but if we can simply remove it that's a much better outcome.

Comment (surayya-MS) [reply src/Components/Components/src/PublicAPI.Unshipped.txt]: > There might still be ways to achieve it (by hand-writing the logic that applies parameters in SectionContent's SetParametersAsync instead of auto-assigning them based on [Parameter] attributes)

The challenge was to then apply it to RenderTreeBuilder here https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web/src/Head/HeadOutlet.cs#L48. Couldn't figure out how to do it.

It a good thing we can remove it altogether.

**Pull request merged**
